File: Data Structures & Algorithms/course-schedule/submission-1.java
class Solution {

    private final Map<Integer, List<Integer>> adjList = new HashMap<>();

    public boolean canFinish(int numCourses, int[][] prerequisites) {

        // create adjcacency list
        for (int[] prerequisite : prerequisites) {
            int course = prerequisite[0];
            int requirement = prerequisite[1];
            if (adjList.containsKey(course)) {
                adjList.get(course).add(requirement);
            } else {
                List<Integer> list = new ArrayList<>();
                list.add(requirement);
                adjList.put(course, list);
            }
        }

        for (int i = 0; i < numCourses; i++) {
            if (dfs(new HashSet<>(), i)) return false;
        }

        return true;
    }

    public boolean dfs(Set<Integer> visited, int current) {

        if (visited.contains(current)) {
            return true;
        }

        List<Integer> children = adjList.get(current);
        if (children == null) {
            return false;
        }

        visited.add(current);
        for (int requirement : children) {
            if (dfs(visited, requirement)) return true;
        }
        visited.remove(current);

        return false;
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        return dfs(nums, 0, -1);
    }

    int dfs(vector<int>& nums, int current, int prev) {
        if (current == nums.size()) return 0;
        int max = dfs(nums, current + 1, prev);
        if (prev == -1 || nums[prev] < nums[current]) {
            max = std::max(max, 1 + dfs(nums, current + 1, current));
        }
        return max;
    }
};
